container-registry-configuration/README.md

```diff
@@ -0,0 +1,189 @@
+
+You are able to define alternative source for an image with a Custom Resource.
+
+```
+apiVersion: saffire.fairwinds.com/v1alpha1
+kind: AlternateImageSource
+metadata:
+  name: alternateimagesource-sample
+spec:
+  imageSourceReplacements:
+    - equivalentRepositories:
+        - quay.io/gianswarm/our-image
+        - azurecr.io/giantswarm/our-image
+        - docker.io/giantswarm/our-image
+```
+
+This mechanism doesn't work for static pods and it doesn't prevent the issue in advance. Also, it has no built-in mechanism for authentication. We can extend it or combine with the image pull secret solution but it sounds too complicated and dirty.
+
+
+#### 1.c Containerd Configuration
+
+As we do in vintage clusters, we can configure containerd configuration as below. This seems the best approach.
+
+```
+[plugins."io.containerd.grpc.v1.cri".registry]
+
+[plugins."io.containerd.grpc.v1.cri".registry.mirrors]
+[plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
+endpoint = ["giantswarm.azurecr.io"]
+
+[plugins."io.containerd.grpc.v1.cri".registry.configs]
+[plugins."io.containerd.grpc.v1.cri".registry.configs."registry-1.docker.io".auth]
+username = "giantswarm-user-account"
+password = "my_token_from_docker_hub"
+```
+
+#### Decision
+
+`1.c` is selected.
+We will use containerd configuration. 
+
+### 2. Authentication
+
+#### 2.a Using unauthenticated accounts
+
+Using only public images without authentication is an option to avoid complexity of securing credentials. However, registry providers have pull/rate limits per account. We will be limited by total of [free account limit](https://docs.docker.com/docker-hub/download-rate-limit/) of Docker Hub (100 pulls per 6 hours per IP address) and [tier limit](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-skus) of Azure Container Registry (3000 ReadOps per minute for standard). Anyone can pull images from our registry and spend our limits, which makes us vulnerable. 
+
+#### 2.b Using authenticated accounts
+
+We can pass credentials to `containerd` configuration to be not limited by registry provider limits.
+
+#### Decision
+
+`2.b` is selected.
+We will use authenticated accounts.
+
+
+### 3. Registries
+
+The infranet page (See `Registry Mirrors`) states this as a desired configuration:
+  - `docker.io` as primary public registry since it has a privilege in docker daemon.
+  - Our own Azure Container Registry as secondary registry because of the security concerns
+
+Since we use `containerd`, docker has not a privelege anymore but there is no need to change this order at the moment. We will follow this configuration for CAPI clusters too.
+
+### 4. Accounts
+
+- 4.a Giantswarm Account + Single account
+- 4.b Giantswarm Account + Per MC including its WCs
+- 4.c Giantswarm Account + Per WC
+- 4.b Customer Accounts
+
+#### Decision
+
+`4.b` is selected.
```

Comment:
When was this decided? My understanding was we were going to use the same approach we do now with vintage and use out own read-only credentials.

Customers can then add their own using `imagePullSecrets`

Reply:
That is my understanding from the discussions. I may be wrong.

What do you plan to do 
- for static pods of WCs like api-server?
- for our non-static pods in WCs like csi-controller-plugin?

Reply:
I still think we should be setting the containerd config but it should be static using our credentials. No need to expose it to customer. There's no requests for it and no reason to expose a potential way for customers to destroy their cluster.

Reply:
This section is not related to whether customers will see our credentials or not. Section 7 is related to that topic. 

Regarding this section, when we put the credentials to containerd configuration of WC nodes, it will be used for all pods of WC. Are you OK with it?

Are you OK with "per MC" approach in account usage? There will be like `giantswarm-hyperion` account and it will be used in hyperion and it's all WCs.

Reply:
That's a change from how we do it in vintage.

In vintage we have a single set of credentials used _everywhere_ to handle authenticated pulls. Those credentials don't have access to anything secret and don't have write permission. Those credentials are baked into the node images and not available to be configured by the customer anywhere.

I thought for this specific issue (avoiding rate limiting) we would be following the same approach. 

I don't see any need to over complicate things with multiple credentials, multiple levels of config, out-of-sync config due to releases not being deployed everywhere yet, etc. It'll make debugging harder for (in my opinion) no benefit.

Reply:
Do you propose embedding credentials to node images in the image builder? If we use the same credentials everywhere, a customer's usage will affect other customers since there are still limits for premium accounts. Shouldn't we isolate tenants?

Reply:
Any idea how vintage avoids that today? Is it an issue we're already seeing?

Reply:
we currently don't isolate between tenants. this is possibly something we should start doing

we'd have to investigate the docker account setup a bit more, i forget the details here

Reply:
iirc, there's some flexibility here, i doubt docker care if we create a bunch of accounts and link them all to our plan

Reply:
let's not mix the baking into node topics with the per MC or not per MC topic.

on the account topic: we currently use one account for all MCs, for Docker it would be the same if we stay that way vs multiple accounts. only benefit of per MC accounts would be that if they get expensive we can make those costs more transparent to the customers, which might be cool, ideally customer should not really re-use our accounts to be honest.

mid term: we should have harbor in the MC and cache/proxy all images through there, which would also significantly reduce requests.

Reply:
We don't need to bake them into the image, they can be part of the cluster app and pulling a secret from the config repo (or wherever)

Reply:
I think we should get into the habit of isolating the MCs from each other (especially with the push for CAPA & CAPZ private clusters) and this includes having separate secrets per MC, even for the registry. I know in this case it's not a big deal because we don't have anything secret in the registry, but a decision like `MCs are isolated from each other and centralized features are an exception` would have avoided the conversation entirely.

I also think the registry and its configuration needs to be externalized (not backed in an image or hard-coded) in order to keep our product flexible. This would allow us to quickly respond to customer requests, debug problems & test features.